### ISSUE TITLE
Simplify the construction of objects from JSON strings

### ISSUE THREAD

**Issue by buddha314** [chapel-lang/chapel#8278]
There is a veritable wave of JSON issues.  In this [Stack Overflow](https://stackoverflow.com/questions/45617341/how-to-populate-a-chapel-class-object-from-a-json) we see you can create objects from JSON string, but the interface is really weird.  WTF is `openmem()` doing in there?

Also, populating arrays is problematic according to https://github.com/chapel-lang/chapel/issues/7582, https://github.com/chapel-lang/chapel/issues/8252 and https://github.com/chapel-lang/chapel/pull/8261 

We are about to start building a REST interface and the common DTO is JSON.  A great first step would be some kind of interface like the SO post, with the Python equivalent

```python
class Fighter {
  var subclass:string;
  var level:int;
}
s = "{'subclass':'Ninja', 'level':7}"
n = json.loads(s)
```

Or the ability to do 

```chapel
var Gimli = new Fighter(s);  // using JSON in the constructor somehow
```

**Comment by nimitbhardwaj** [comment]
Yes, I also agree with this statement, as I tried to work on the #8261, there I see there is no modular approach used to write/read the JSON output/input using writef and readf functions, as in the code, at some point in code I read how readf and writef understands json, it do it empirically, its not modular now, like that function is just handling JSON and non JSON formats. So I think making JSON objects is fine as compared to just reading the memory where the variable is stored and interpreting it later.

**Comment by mppf** [comment]
Yes it's a known issue that we need some JSON support outside of writef/readf. See also issue #7846 which proposes some kind of "formatter" that could interact with writef/readf. Presumably you'd be able to make calls directly on such an object if you preferred not to use writef/readf.

I think that the TOML module https://github.com/chapel-lang/chapel/blob/master/modules/packages/TOML.chpl is an example of a module that provides read/write facilities for a particular format. I think it'd be reasonable to create something like it for JSON. In that event, instead of reading Chapel types natively, reading some JSON would create JSONArray or JSONObject types.

> What is openmem() doing in there?

To answer your question, some of the I/O oddities come about because I assumed that we'd be able to create a connection between string and I/O files/channels... so you'd be able to use a string anywhere you can use a file/channel. That hasn't come about yet, and using openmem is the workaround.

**Comment by mppf** [comment]
> WTF

I'd really prefer we leave off the "WTF" part; I interpret it as a value judgement on my work that I don't appreciate. I can certainly understand frustration as you try to use it. Of course, one can communicate confusion and frustration without being potentially insulting. I happen to think that openmem is a cool and powerful feature.

**Comment by nimitbhardwaj** [comment]
Hey, there is one more thing I want to ask for this issue, its that, in internal/ChapelIO.chpl, in line 279, there is __primitive using field by num, as one argument, it automatically gives string as quoted with " and ", but for others such as int and (for now enum) it don't encode with " and ", and also for non JSON formats it don't enclose them, I am not able to get it why its doing so, but no checks are performed there (or may be I didn't get em).
That function was returning the read or write string in json or non-json form.

**Comment by mppf** [comment]
@nimitbhardwaj - that primitive just gets a reference to each field (one at a time). The writer.readwrite call is the one doing the actual I/O. writer is expected to be a channel, and channel.readwrite is defined in IO.chpl and it eventually calls _read_one_internal or _write_one_internal, which for the particular case you're interested in calls _read_text_internal / _write_text_internal - (and you already know that code can be adjusted for enums).

**Comment by nimitbhardwaj** [comment]
Yes, I was thinking same, i was right, there it can be improved, very very thank you sir.

**Comment by lydia-duncan** [comment]
@buddha314 - note that @DanilaFe resolved this with his #25464.  Thanks for reporting, and thanks for the fix, Daniel!
